packages/next/src/server/app-render/app-render.tsx

```diff
@@ -138,11 +138,13 @@ function findDynamicParamFromRouterState(
 export async function renderToHTMLOrFlight(
   req: IncomingMessage,
   res: ServerResponse,
-  pathname: string,
+  pagePath: string,
   query: NextParsedUrlQuery,
   renderOpts: RenderOpts
 ): Promise<RenderResult> {
   const isFlight = req.headers[RSC.toLowerCase()] !== undefined
+  const initialCanonicalUrl = validateURL(req.url)
+  const pathname = new URL(req.url!, 'http://n').pathname
```

Comment:
I remember there being a way to provide inputs that would bypass this and create an entirely different url when used on user input.

Reply:
Fixed in [8502841](https://github.com/vercel/next.js/pull/49512/commits/85028419cf011154eb614e4c0460b94d3a7856aa)